File: backend/app/models/product.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import Column, String, Text, Float, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from app.core.database import Base

class Product(Base):
    __tablename__ = "products"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    artisan_id = Column(String(36), ForeignKey("artisans.id", ondelete="CASCADE"), nullable=False, index=True)
    title = Column(String(200), nullable=False, index=True)
    slug = Column(String(220), unique=True, nullable=False, index=True)
    description_en = Column(Text, nullable=False)
    description_hi = Column(Text, nullable=True)
    category_id = Column(String(36), ForeignKey("categories.id", ondelete="RESTRICT"), nullable=False, index=True)
    subcategory_id = Column(String(36), ForeignKey("subcategories.id", ondelete="SET NULL"), nullable=True)
    tradition_id = Column(String(36), ForeignKey("traditions.id", ondelete="RESTRICT"), nullable=False, index=True)
    
    # Transparent Price Breakdown Fields (in INR)
    base_price = Column(Float, nullable=False) # Artisan manufacturing/material price
    artisan_share = Column(Float, nullable=False) # Total artisan earning from sale
    platform_fee = Column(Float, default=0.0, nullable=False) # Platform commission & maintenance
    delivery_fee = Column(Float, default=0.0, nullable=False) # Logistics
    
    ai_confidence_score = Column(Float, nullable=True) # Classification confidence score
    certificate_id = Column(String(100), unique=True, nullable=True, index=True)
    certificate_hash = Column(String(64), nullable=True) # SHA-256 immutable digest
    
    status = Column(String(30), default="draft", nullable=False, index=True) # draft, pending_review, published, rejected
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    artisan = relationship("Artisan", lazy="selectin")
    category = relationship("Category", lazy="selectin")
    subcategory = relationship("Subcategory", lazy="selectin")
    tradition = relationship("Tradition", lazy="selectin")
    images = relationship("ProductImage", back_populates="product", cascade="all, delete-orphan", lazy="selectin")
    variants = relationship("ProductVariant", back_populates="product", cascade="all, delete-orphan", lazy="selectin")
    moderation_logs = relationship("ProductModerationLog", back_populates="product", cascade="all, delete-orphan", lazy="selectin")
    certification = relationship("ProductCertification", back_populates="product", uselist=False, cascade="all, delete-orphan", lazy="selectin")

    @property
    def total_price(self) -> float:
        return round(self.base_price + self.platform_fee + self.delivery_fee, 2)

File: packages/shared-types/product.ts
export type ProductStatus = 'draft' | 'pending_review' | 'published' | 'rejected';

export interface PriceBreakdown {
  base_price: number;
  artisan_share: number;
  platform_fee: number;
  delivery_fee: number;
  tax: number;
  total_price: number;
}

export interface ProductVariant {
  id: string;
  product_id: string;
  variant_name: string;
  price_delta: number;
  stock_quantity: number;
}

export interface ProductImage {
  id: string;
  product_id: string;
  image_url: string;
  display_order: number;
  is_primary: boolean;
}

export interface ProductCertification {
  id: string;
  product_id: string;
  certificate_id: string;
  certificate_hash: string;
  qr_code_url: string;
  craft_tradition: string;
  artisan_name: string;
  origin_region: string;
  issued_at: string;
}

export interface Product {
  id: string;
  artisan_id: string;
  artisan?: {
    id: string;
    display_name: string;
    region: string;
    avg_rating: number;
    verification_status: string;
  };
  title: string;
  description_en: string;
  description_hi?: string;
  category_id: string;
  subcategory_id?: string;
  tradition_id: string;
  pricing: PriceBreakdown;
  ai_confidence_score?: number;
  status: ProductStatus;
  images: ProductImage[];
  variants: ProductVariant[];
  certification?: ProductCertification;
  created_at: string;
}
